src/ngraph/pass/concat_fusion.cpp

```diff
@@ -0,0 +1,304 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "concat_fusion.hpp"
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include <unordered_set>
+
+#include "ngraph/graph_util.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/broadcast.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/parameter.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/pattern/op/label.hpp"
+#include "ngraph/pattern/op/skip.hpp"
+#include "ngraph/util.hpp"
+
+using namespace ngraph;
+
+bool check_self_concat_op(const std::shared_ptr<Node>& op)
+{
+    auto input_args = op->get_arguments();
+    std::set<std::shared_ptr<Node>> input_args_set(input_args.begin(), input_args.end());
+    return (input_args_set.size() == 1);
+}
+
+bool check_concat_axis_dim_value(const std::shared_ptr<Node>& concat_op)
+{
+    auto input_shape = concat_op->get_input_shape(0);
+    size_t concat_axis = std::static_pointer_cast<op::Concat>(concat_op)->get_concatenation_axis();
+
+    return (input_shape[concat_axis] == 1);
+}
+
+bool check_concat_has_no_fan_out(const std::shared_ptr<Node>& op)
+{
+    auto users = op->get_users();
+    std::set<std::shared_ptr<Node>> user_set(users.begin(), users.end());
+    size_t num_unique_users = user_set.size();
+    if (num_unique_users == 1)
+    {
+        return true;
+    }
+    else
+    {
+        NGRAPH_DEBUG << "self_concat_fusion: " << op->get_name() << " has fan out\n";
+        return false;
+    }
+}
+
+bool valid_self_concat(const std::shared_ptr<Node>& Op)
+{
+    if (!check_self_concat_op(Op))
+    {
+        NGRAPH_DEBUG << "self_concat_fusion: Matcher matched " << Op->get_name()
+                     << " but it is not a self concat\n";
+        return false;
+    }
+
+    if (!check_concat_axis_dim_value(Op))
+    {
+        NGRAPH_DEBUG << "self_concat_fusion: Input shape value along concat axis of "
+                     << Op->get_name() << " is not equal to 1\n";
+        return false;
+    }
+
+    return true;
+}
+
+void pass::ConcatElimination::construct_concat_elimination()
+{
+    auto op_label = std::make_shared<pattern::op::Label>(element::f32, Shape{1, 3});
+    auto concat = std::make_shared<op::Concat>(NodeVector{op_label}, 0);
+    auto concat_label = std::make_shared<pattern::op::Label>(concat, nullptr, NodeVector{concat});
+
+    auto callback = [op_label](pattern::Matcher& m) {
+        NGRAPH_DEBUG
+            << "concat_elimination: In callback for construct_concat_elimination against node = "
+            << m.get_match_root()->get_name();
+        auto pattern_map = m.get_pattern_map();
+        auto op = pattern_map[op_label];
+
+        auto root = std::dynamic_pointer_cast<op::Concat>(m.get_match_root());
+        if (root && (root->get_input_shape(0) == root->get_output_shape(0)))
```

Comment:
The input shape and output shape are the same for `Concat`?

Reply:
Yes..for elimination of a concat in this case, the shapes will be same...as in the concat does nothing. Actually this check is redundant but I put it in to drive home the point. Technically, any concat that has 1 input must have the same input and output shape.